midp/gfx.js

```diff
@@ -98,10 +169,20 @@ var currentlyFocusedTextEditor;
 
     var refreshStr = "refresh";
     Native["com/sun/midp/lcdui/DisplayDevice.refresh0.(IIIIII)V"] = function(hardwareId, displayId, x1, y1, x2, y2) {
-        x1 = Math.max(0, x1);
-        y1 = Math.max(0, y1);
-        x2 = Math.max(0, x2);
-        y2 = Math.max(0, y2);
+        if (!config.useOffscreenCanvas) {
+            // If the offscreen canvas is disabled, then we draw directly
+            // to the device canvas, so it's always up-to-date and never needs
+            // to be refreshed.
+            return;
+        }
+
+        x1 = Math.max(0, x1) * DPR;
+        y1 = Math.max(0, y1) * DPR;
+        x2 = Math.max(0, x2) * DPR;
+        y2 = Math.max(0, y2) * DPR;
+
+        // XXX Isn't the offscreen and device canvas width and height always
+        // the same, so it isn't necessary to compute which one is more minimal?
```

Comment:
Should be. If there are edge cases that cause us to need to do this min check (e.g. maybe when we rotate the screen during an animation that has repaint events pending) we should probably document those explicitly here.